Support for Kinesis

justinsb opened the pull request:
Basic support for Kinesis

Depends on PR #2

justinsb committed:
Add kinesis binding & tests

sclasen commented:
@justinsb Looks great!!  thanks for these.

 Two questions/asks
- can we use a `com.amazonaws.transform.VoidJsonUnmarshaller<T>` explicitly, rather than passing null as the implicit unmarshaller? might make it slighty clearer as to what is happening? 
- can you map the Future[Void]s to Future[Unit]s  so that the clent interface is more scala-rific?

Neither are a huge deal. Thoughts?

justinsb committed:
Fix isDefined -> isEmpty

sclasen commented:
@justinsb ping on comments above?

justinsb commented:
Absolutely agree with both suggestions - new to Scala, so these hints are very welcome.  Sorry for delay!  I'll try to update the PR today.

sclasen commented:
:+1:

justinsb committed:
Create unmarshaller for Scala Unit results

justinsb committed:
Adapt tests to today's Kinesis behaviour: delay and fewer getRecords iterations

Last week, it was delaying returning records in getRecords,
so we had to loop through a large number.

Today, it is returning records immediately, but enforcing the throughput
limit which it wasn't before.  So we can't loop through so many records.

Compromise: reduce the number of loop iterations from 128 to 32, and insert
a pause of 50 ms in between getRecord calls.  That should cover 'reasonable'
behaviour.

justinsb commented:
Updated with a direct-to-Unit unmarshaller (seemed the easiest way).

Also Kinesis is behaving a bit differently today, so I had to adapt the integration tests.  Thinking about it more, I just re-created the stream, so I bet it behaves differently on new vs older streams.  Anyway, the behaviour we're testing now seems reasonable: it should avoid throttling, and if we have to do more than 32 iterations through getRecords, we would probably like to know that.

sclasen commented on spray-kinesis/src/main/scala/com/sclasen/spray/aws/kinesis/KinesisClient.scala:
```diff
@@ -0,0 +1,139 @@
+ * Like VoidJsonUnmarshaller, but returns a Scala Unit rather than a Java Void/null.
+ */
+private object UnitUnmarshaller extends Unmarshaller[Unit, JsonUnmarshallerContext] {
+  def unmarshall(context: JsonUnmarshallerContext): Unit = {
```
this can be written as `def unmarshall(context: JsonUnmarshallerContext): Unit = ()`

sclasen commented:
@justinsb thanks! one line comment above plus I pushed a change to master that will break this PR 

see https://github.com/sclasen/spray-aws/blob/master/spray-aws/src/main/scala/com/sclasen/spray/aws/SprayAWSClient.scala#L101

the response function now returns a `Either[AmazonServiceException, T]`  so we can differentiate between 3 states.  Right(T) means the request succeeded Left(AmazonServiceException) means AWS returned an exception, and when the Future fails we had some client side issue with spray etc...

Take a look at how I changed the dynamo client. Still has the send\* methods that return Future[T] but they are implemented by calling fold on the corresponding method that returns Future[Either[AmazonServiceException,T]] like so...

```
 def batchGetItem(awsGet: BatchGetItemRequest): Future[Either[AmazonServiceException, BatchGetItemResult]] =
    pipeline(request(awsGet)).map(response[BatchGetItemResult])

  def sendDeleteItem(awsDel: DeleteItemRequest): Future[DeleteItemResult] = fold(deleteItem(awsDel))
```

Im happy to merge this as is and fix it up myself, but if you want to do it that would work too. Just rebase this branch on current master and force push it back after updating the kinesis client.

Let me know if you want me to merge or if you want to fix it up.

justinsb commented:
I'll do it :-)

justinsb committed:
Use more idiomatic scala

justinsb committed:
Use new API: X returns a Future with 'Either', sendX folds it to be a raw future

sclasen commented:
@justinsb hey didnt see that you pushed up the Either stuff!  Thanks.

I merged another PR that breaks this one, so I will fix it up and push it to master

justinsb commented:
Awesome - thanks.

Let me know if you need anything from me.
